File: packages/cast/android/src/main/java/com/margelo/nitro/rnmediacast/HybridRnMediaCastButton.kt
// Same hard rule as `HybridRnMediaCast.kt`: nitrogen hardcodes the
// implementation class into the generated ViewManager as
// `com.margelo.nitro.<androidNamespace>.<implementationClassName>` (see
// nitrogen/generated/android/kotlin/.../views/HybridRnMediaCastButtonManager.kt),
// so this file cannot move to `com.rnmediacast`.
package com.margelo.nitro.rnmediacast

import android.content.Context
import android.graphics.Canvas
import android.graphics.PorterDuff
import android.view.View
import androidx.mediarouter.app.MediaRouteButton
import com.facebook.react.uimanager.ThemedReactContext
import com.google.android.gms.cast.framework.CastButtonFactory

/**
 * The Android half of `<CastButton/>`: the framework's own cast affordance,
 * wrapped as a Nitro HybridView.
 *
 * Why `MediaRouteButton` + `CastButtonFactory` and not our own `Pressable`
 * calling `showCastPicker()` — the whole reason this component exists:
 * `CastButtonFactory.setUpMediaRouteButton(Context, MediaRouteButton)` is the
 * only wiring that reaches the **system output switcher**. Verified in the
 * shipped artifacts (play-services-cast-framework 22.3.1,
 * mediarouter 1.8.0-beta01):
 *
 * - `CastButtonFactory.setUpMediaRouteButton` sets the button's selector from
 *   `CastContext.getMergedSelector()` and registers it so a later
 *   `CastContext` init re-applies it. It tolerates a GMS-less device: the
 *   internal `CastContext` getter catches the load failure and returns null
 *   ("Failed to load module from Google Play services … Ignoring this failure
 *   silently"), so construction never throws.
 * - `MediaRouteButton.showDialog()` then branches on
 *   `MediaRouterParams.isOutputSwitcherEnabled() && MediaRouter.isMediaTransferEnabled()`
 *   → `SystemOutputSwitcherDialogController.showDialog(context)` (the SYSTEM
 *   sheet, Android 13+ with `MediaTransferReceiver` in the manifest), else the
 *   in-app `MediaRouteChooserDialogFragment`. The first branch is switched on
 *   by `CastOptions.setShowSystemOutputSwitcherOnCastIconClick(true)`, which
 *   `RnMediaCastOptionsProvider` already sets.
 *
 * Threading contract
 * ------------------
 * UI THREAD ONLY, and that is guaranteed rather than defended: React Native
 * creates views from `SurfaceMountingManager.createView`, which is
 * `@UiThread`, and pushes props from the ViewManager's `updateState` on the
 * same thread. Both `MediaRouter.getInstance` (in the `MediaRouteButton`
 * constructor) and `CastButtonFactory.setUpMediaRouteButton`
 * (`Preconditions.checkMainThread("Must be called from the main thread.")`)
 * assert it themselves, so a violation would be loud, never silent.
 *
 * Deliberately NOT a `RecyclableView`: a recycled button would keep the
 * `MediaRouteButton` bound to the Activity of its first surface, and that
 * Activity is what the pre-Android-13 chooser dialog is shown from
 * (`MediaRouteButton.getActivity()` unwraps the view's context). A fresh
 * button per mount costs one view allocation and cannot show a dialog on a
 * dead Activity.
 */
class HybridRnMediaCastButton(context: ThemedReactContext) :
  HybridRnMediaCastButtonSpec() {

  private val button = TintedMediaRouteButton(context)

  override val view: View = button

  override var tintColor: Double? = null
    set(value) {
      field = value
      // `processColor` packs ARGB into a (signed, on Android) 32-bit int and
      // Nitro carries it as a double; `toInt()` is the exact inverse for
      // every value it can produce.
      button.iconTint = value?.toInt()
    }

  init {
    CastButtonFactory.setUpMediaRouteButton(context, button)
  }
}

/**
 * A [MediaRouteButton] whose icon can be recoloured per instance.
 *
 * `MediaRouteButton` takes its icon tint from the *theme* attribute
 * `mediaRouteButtonTint`, read once in its constructor; there is no public
 * per-instance setter, and `setRemoteIndicatorDrawable` would mean shipping
 * our own copy of Google's cast icon (and losing the connecting animation and
 * the connected state that come with theirs — the Cast Design Checklist wants
 * the standard icon).
 *
 * So the icon is recoloured where it is drawn: `super.onDraw` renders into an
 * offscreen layer, then `SRC_IN` paints the tint through exactly the pixels it
 * covered. Whatever state the framework draws — idle, connecting animation,
 * connected — keeps its shape and gets the app's colour. The layer is only
 * allocated when a tint is actually set, and a cast button repaints on route
 * changes, not per frame.
 */
private class TintedMediaRouteButton(context: Context) : MediaRouteButton(context) {
  /** Packed ARGB, or null for "leave the framework's own tint alone". */
  var iconTint: Int? = null
    set(value) {
      if (field == value) return
      field = value
      invalidate()
    }

  override fun onDraw(canvas: Canvas) {
    val tint = iconTint
    if (tint == null) {
      super.onDraw(canvas)
      return
    }
    val layer = canvas.saveLayer(null, null)
    super.onDraw(canvas)
    canvas.drawColor(tint, PorterDuff.Mode.SRC_IN)
    canvas.restoreToCount(layer)
  }
}

File: packages/cast/ios/HybridRnMediaCastButton.swift
//
//  HybridRnMediaCastButton.swift
//  RnMediaCast
//
//  The iOS half of `<CastButton/>`: Google's own cast button, wrapped as a
//  Nitro HybridView.
//
//  Why the SDK's button and not a `Pressable` calling `showCastPicker()`:
//  `GCKUICastButton` is a `UIButton` subclass (GoogleCast/GCKUICastButton.h:22,
//  umbrella-exported at GoogleCast/GoogleCast.h:68) that owns the cast icon's
//  three states — inactive, connecting animation, connected — and, with its
//  default `triggersDefaultCastDialog == YES` (GCKUICastButton.h:33), presents
//  the framework's own device dialog on tap. That default is deliberately left
//  alone here: it is also the SDK's local-network-permission choreography.
//  Discovery starts on the FIRST TAP of a cast button, never before, so iOS
//  shows its local-network prompt exactly when the user asked for devices —
//  the ordering Google's own checklist requires and the one an app-drawn
//  button gets wrong.
//
//  Threading contract
//  ------------------
//  MAIN THREAD ONLY, guaranteed by the caller rather than defended here: the
//  generated `HybridRnMediaCastButtonComponent.mm` creates this object from
//  `-[RCTViewComponentView init]` and pushes props from `updateProps:oldProps:`,
//  both of which are UIKit-thread only. GoogleCast and UIKit both require it.
//

import Foundation
import GoogleCast
import NitroModules
import UIKit

final class HybridRnMediaCastButton: HybridRnMediaCastButtonSpec {
  /// The SDK's button, sized by React Native's layout like any other view.
  ///
  /// A Nitro view's shadow node is a plain `ConcreteViewShadowNode` — it has no
  /// measure function, so the button has no intrinsic size in the RN tree and
  /// takes exactly the frame `style` gives it. The icon itself is drawn at its
  /// own size, centred; the view's size is the touch target. The JS component
  /// therefore ships a default width/height (see `cast-button.tsx`).
  let view = GCKUICastButton(frame: .zero)

  /// Packed ARGB from `processColor`, or `nil` for "inherit".
  var tintColor: Double? {
    didSet {
      // `nil` restores the inherited tint rather than forcing a colour —
      // `UIView.tintColor = nil` is documented as "use the superview's".
      view.tintColor = tintColor.flatMap(HybridRnMediaCastButton.color(fromPackedARGB:))
    }
  }

  /// `processColor` packs a colour into 0xAARRGGBB (signed on Android, which
  /// is why the negative case is handled) and Nitro carries it as a `Double`.
  private static func color(fromPackedARGB packed: Double) -> UIColor? {
    guard packed.isFinite else { return nil }
    let argb = UInt32(truncatingIfNeeded: Int64(packed))
    return UIColor(
      red: CGFloat((argb >> 16) & 0xFF) / 255,
      green: CGFloat((argb >> 8) & 0xFF) / 255,
      blue: CGFloat(argb & 0xFF) / 255,
      alpha: CGFloat((argb >> 24) & 0xFF) / 255
    )
  }
}
